Sources/Testing/Running/EntryPoint.swift

```diff
@@ -244,6 +260,75 @@ func runTests(options: [Event.ConsoleOutputRecorder.Option], configuration: Conf
   await runner.run()
 }
 
+// MARK: - Experimental event streaming
+
+#if !SWT_NO_FILE_IO && canImport(Foundation)
+/// A type containing an event snapshot and snapshots of the contents of an
+/// event context suitable for streaming over JSON.
+///
+/// This function is not part of the public interface of the testing library.
+/// External adopters are not necessarily written in Swift and are expected to
+/// decode the JSON produced for this type in implementation-specific ways.
+struct EventAndContextSnapshot {
+  /// A snapshot of the event.
+  var event: Event.Snapshot
+
+  /// A snapshot of the test associated with the event, if any.
+  var test: Test.Snapshot?
+
+  /// A snapshot of the test case associated with the event, if any.
+  var testCase: Test.Case.Snapshot?
+}
+
+extension EventAndContextSnapshot: Codable {}
+
+/// Create an event handler that streams events to the file at a given path.
+///
+/// - Parameters:
+///   - path: The path to which events should be streamed. This file will be
+///     opened for writing.
+///
+/// - Throws: Any error that occurs opening `path`. Once `path` is opened,
+///   errors that may occur writing to it are handled by the resulting event
+///   handler.
+///
+/// - Returns: An event handler.
+///
+/// For each event handled by the resulting event handler, a JSON blob
+/// representing it and its associated context is created. The length of the
+/// blob, as a 64-bit little-endian integer, is written to `path`, followed by
+/// the blob itself.
+///
+/// The file at `path` can be a regular file, however to allow for streaming a
+/// named pipe is recommended. `mkfifo()` can be used on Darwin and Linux to
+/// create a named pipe; `CreateNamedPipeA()` can be used on Windows.
+///
+/// The file at `path` is closed when this process terminates or the
+/// corresponding call to ``Runner/run()`` returns, whichever occurs first.
+private func _eventHandlerForStreamingEvents(toFileAtPath path: String) throws -> Event.Handler {
+  // Open the event stream file for writing.
+  let file = try FileHandle(forWritingAtPath: path)
+
+  return { event, context in
+    let snapshot = EventAndContextSnapshot(
+      event: Event.Snapshot(snapshotting: event),
+      test: context.test.map { Test.Snapshot(snapshotting: $0) },
+      testCase: context.testCase.map { Test.Case.Snapshot(snapshotting: $0) }
+    )
+    if let snapshotJSON = try? JSONEncoder().encode(snapshot) {
+      try? file.withLock {
```

Comment:
For these `try?` usages could we consider some fallback logic which encodes an explanation of the error, or something to help convey to a supporting tool that an error occurred and data is missing in the stream?

If not for this PR, then maybe a "FIXME" comment to note it for later improvement.

Reply:
For the JSON failure, no, because the `try?` can't really fail for these types, so it'd just be dead code. And for the file calls, if they throw an error, that means there's nowhere to write the diagnostic data.

Reply:
Why is it not possible for the JSON encoding to potentially fail for these types?

Reply:
Because the types we're encoding have no failure modes.

Reply:
Certainly would not consider this specific thing a blocker, but I do think we ought to eventually incorporate some way for this stream to convey errors which occurred during a test run. This is conceptually different than conveying `Issue`s recorded _by_ tests—what I'm describing is more akin to an infrastructure failure. 

That's the sort of scenario I had in mind here: what if our `Snapshot` types fail to encode due to a problem with their contents, such as un-serializable data values supplied at runtime by a test? Or more generally: if something about running tests goes wrong, how can we represent that in this stream?

Reply:
It's worth noting that the `Runner.run()` function is not `throw`ing, and we do have `Issue.Kind.system`, which is intended to represent that sort of problem when recorded with no associated test or case. We may still conclude that it would be useful to have some additional category of error reporting possible, but I'm not sure we have enough data on hand right now to design it properly.

Reply:
Perhaps we ought to just crash upon an encoding failure, though? It’d be a very nasty bug to investigate if we ended up silently dropping events in the stream due to some future regression in the encoding implementation.

Reply:
Crashing the test process is a catastrophic failure and we only ever reach for `fatalError()` and friends when there's a bug in our code that prevents forward progress. It seems heavy-handed to crash the process because a pipe got closed early or something.

Reply:
Oh, I was suggesting that only for a failure from JSONEncoder, not for a file I/O problem.

Reply:
Consider also: it seems to me that a dropped event represents data loss in this context, and silent data loss seems worse than noisy data loss?

Reply:
If data is being lost silently, then our only options are to remain silent or to cause _more_ data loss with early termination.